src/debug/validatePreDebug.ts

```diff
@@ -0,0 +1,88 @@
+    shouldContinue: boolean;
+}
+
+export async function preDebugValidate(debugConfig: vscode.DebugConfiguration): Promise<IPreDebugValidateResult> {
+    const workspace: vscode.WorkspaceFolder = getMatchingWorkspace(debugConfig);
+    let shouldContinue: boolean = await validateFuncCoreToolsInstalled();
+    if (shouldContinue) {
+        const projectPath: string | undefined = await tryGetFunctionProjectRoot(workspace.uri.fsPath, true /* suppressPrompt */);
+        if (projectPath) {
+            shouldContinue = await validateEmulatorIsRunning(projectPath);
+        }
+    }
+
+    return { workspace, shouldContinue };
+}
+
+function getMatchingWorkspace(debugConfig: vscode.DebugConfiguration): vscode.WorkspaceFolder {
+    if (vscode.workspace.workspaceFolders) {
+        for (const workspace of vscode.workspace.workspaceFolders) {
+            try {
+                const configs: vscode.DebugConfiguration[] = getDebugConfigs(workspace);
+                if (configs.some(c => isDebugConfigEqual(c, debugConfig))) {
+                    return workspace;
+                }
+            } catch {
+                // ignore and try next workspace
+            }
+        }
+    }
+
+    throw new Error(localize('noDebug', 'Failed to find launch config matching name "{0}", request "{1}", and type "{2}".', debugConfig.name, debugConfig.request, debugConfig.type));
+}
+
+/**
+ * If AzureWebJobsStorage is set, pings the emulator to make sure it's actually running
+ */
+async function validateEmulatorIsRunning(projectPath: string): Promise<boolean> {
+    const azureWebJobsStorage: string | undefined = await getAzureWebJobsStorage(projectPath);
+    if (azureWebJobsStorage && azureWebJobsStorage.toLowerCase() === localEmulatorConnectionString.toLowerCase()) {
+        try {
+            const client: azureStorage.BlobService = azureStorage.createBlobService(azureWebJobsStorage);
+            await new Promise((resolve, reject): void => {
+                // Checking against a common container for functions, but doesn't really matter what call we make here
+                client.doesContainerExist('azure-webjob-hosts', (err: Error | undefined) => {
+                    // tslint:disable-next-line: no-void-expression
+                    err ? reject(err) : resolve();
+                });
+            });
+        } catch (error) {
+            const message: string = localize('failedToConnectEmulator', 'Failed to verify "{0}" connection specified in "{1}". Is the local emulator installed and running?', azureWebJobsStorageKey, localSettingsFileName);
+            const learnMoreLink: string = isWindows ? 'https://aka.ms/AA4ym56' : 'https://aka.ms/AA4yef8';
+
+            const debugAnyway: vscode.MessageItem = { title: localize('debugAnyway', 'Debug anyway') };
+            try {
+                const result: vscode.MessageItem = await ext.ui.showWarningMessage(message, { learnMoreLink, modal: true }, debugAnyway);
```

Comment:
What other error are you expecting here?  It seems like it's just a dialog prompt, so I'm not sure why you have try/catch here.

Reply:
The catch block only swallows UserCancelledErrors - aka I want to return false instead of throwing a UserCancelledError. I can add a comment if that's confusing

Reply:
Eh, I kind of suspected as much, but I wanted to verify since it seemed like a weird pattern to me.  I'm a dummy who is easily confused, but I think a comment would be good.